salt/states/grains.py

```diff
@@ -10,58 +10,81 @@
 Note: This does NOT override any grains set in the minion file.
 '''
 
+from __future__ import absolute_import
+from salt.defaults import DEFAULT_TARGET_DELIM
+import re
 
-def present(name, value):
+
+def present(name, value, delimiter=DEFAULT_TARGET_DELIM, force=False):
     '''
     Ensure that a grain is set
 
+    .. versionchanged:: Boron
+
     name
         The grain name
 
     value
         The value to set on the grain
 
-    If the grain with the given name exists, its value is updated to the new value.
-    If the grain does not yet exist, a new grain is set to the given value.
+    :param force: If force is True, the existing grain will be overwritten
+        regardless of its existing or provided value type. Defaults to False
+        .. versionadded:: Boron
+
+    :param delimiter: A delimiter different from the default can be provided.
+        .. versionadded:: Boron
+
+    It is now capable to set a grain to a complex value (ie. lists and dicts)
+    and supports nested grains as well.
+
+    If the grain does not yet exist, a new grain is set to the given value. For
+    a nested grain, the necessary keys are created if they don't exist. If
+    a given key is an existing value, it will be converted, but an existing value
+    different from the given key will fail the state.
+
+    If the grain with the given name exists, its value is updated to the new
+    value unless its existing or provided value is complex (list or dict). Use
+    `force: True` to overwrite.
 
     .. code-block:: yaml
 
       cheese:
         grains.present:
           - value: edam
+
+      nested_grain_with_complex_value:
+        grains.present:
+          - name: icinga:Apache SSL
+          - value:
+            - command: check_https
+            - params:  -H localhost -p 443 -S
     '''
+    name = re.sub(delimiter, DEFAULT_TARGET_DELIM, name)
     ret = {'name': name,
            'changes': {},
            'result': True,
            'comment': ''}
-    if isinstance(value, dict):
-        ret['result'] = False
-        ret['comment'] = 'Grain value cannot be dict'
-        return ret
-    if __grains__.get(name) == value:
+    existing = __salt__['grains.get'](name, 'nonexistingzniQSDpWy3ywmgFn66lL4PLG9oBrDi')
```

Comment:
I'm thinking if we just use a basic object and use `is` it would be better:

```
unique = object()
existing = __salt__['grains.get'](name, unique)
if existing is unique or existing != value:
    # the grain is not set, set it
```

By using `is`, our new object is guaranteed to not be the same object as any other object. Does that make sense?